rotkehlchen/chain/evm/decoding/decoder.py

```diff
@@ -555,6 +560,78 @@ def run_all_post_decoding_rules(
 
         return decoded_events
 
+    @staticmethod
+    def _process_swaps(
+            transaction: EvmTransaction,
+            decoded_events: list['EvmEvent'],
+    ) -> list['EvmEvent']:
+        """Convert EvmEvents with event_type of TRADE into EvmSwapEvents.
+
+        Assumes that the decoding logic has already ordered the sequence indexes of trade events
+        in the correct spend/receive/fee order with no other events between them (although the
+        indexes do not need to be consecutive). If an incomplete or unordered group of Trade events
+        is encountered an error will be logged and the original EvmEvents saved to the db.
+
+        Returns the list of decoded events ordered by sequence index with any complete groups
+        of trade events replaced with EvmSwapEvents.
+        """
+        processed_events = []
+        trade_subtypes = (HistoryEventSubType.SPEND, HistoryEventSubType.RECEIVE, HistoryEventSubType.FEE)  # noqa: E501
+        events_iterator = peekable(iter(sorted(decoded_events, key=lambda x: x.sequence_index)))
+        while (next_event := events_iterator.peek(None)) is not None:
+            if (
+                next_event.event_type != HistoryEventType.TRADE or
+                next_event.event_subtype not in trade_subtypes
+            ):  # event is not part of a swap - save it and continue
+                processed_events.append(next(events_iterator))
+                continue
+
+            # Temporarily skip counterparties whose tests have not been updated.
+            # TODO: Update all decoding tests involving swaps and remove this counterparty check.
+            if next_event.counterparty not in {CPT_METAMASK_SWAPS, CPT_LLAMAZIP, CPT_PENDLE}:
+                processed_events.append(next(events_iterator))
+                continue
```

Comment:
also, this won't be needed. we'll only need to check that the swap has a spend/receive or spend/receive/fee